### DISCUSSION TITLE
`support` app tries to connect to external endpoint and does not indicate anywhere that it fails to; making `occ` commands to run for 30+ seconds

### DISCUSSION THREAD

**Discussion by dee-kryvenko** [nextcloud/all-in-one#3636] [Questions] [answered]
I got AIO behind reverse proxy on a separate server. AIO would spend almost a full hour to start `nextcloud` container. I was watching the logs and it literally would take 30 to 60 seconds to produce one line of output.

Resources are not a constrain - 16 vCPU, 64GB RAM, SSD...

During my experiments, I found out `php occ` commands basically would took forever to run. There is nothing in the output, no issues anywhere... so I decided to debug a little.

In the `lib/private/Console/Application.php` `loadCommands` loop, I have found out it is getting stuck for ~30 seconds loading `$appPath=/var/www/html/apps/support`. I am not familiar with the code base to dig much deeper... so I decided to look into `strace`.

Many many hours later, what I've found was this:

```
connect(10, {sa_family=AF_INET, sin_port=htons(443), sin_addr=inet_addr("192.168.35.1")}, 16) = -1 EINPROGRESS (Operation in progre
ss) <0.000040>
rt_sigaction(SIGPIPE, {sa_handler=SIG_IGN, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1628d845a7}, NULL, 8) = 0 <
0.000004>
poll([{fd=10, events=POLLOUT}, {fd=7, events=POLLIN}], 2, 0) = 0 (Timeout) <0.000005>
rt_sigaction(SIGPIPE, NULL, {sa_handler=SIG_IGN, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1628d845a7}, 8) = 0 <
0.000004>
rt_sigaction(SIGPIPE, {sa_handler=SIG_IGN, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1628d845a7}, NULL, 8) = 0 <
0.000004>
poll([{fd=10, events=POLLPRI|POLLOUT|POLLWRNORM}], 1, 0) = 0 (Timeout) <0.000004>
rt_sigaction(SIGPIPE, {sa_handler=SIG_IGN, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1628d845a7}, NULL, 8) = 0 <
0.000004>
poll([{fd=10, events=POLLOUT}, {fd=7, events=POLLIN}], 2, 199) = 0 (Timeout) <0.199525>
rt_sigaction(SIGPIPE, NULL, {sa_handler=SIG_IGN, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1628d845a7}, 8) = 0 <
0.000088>
rt_sigaction(SIGPIPE, {sa_handler=SIG_IGN, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1628d845a7}, NULL, 8) = 0 <
0.000076>
poll([{fd=10, events=POLLPRI|POLLOUT|POLLWRNORM}], 1, 0) = 0 (Timeout) <0.000082>
rt_sigaction(SIGPIPE, {sa_handler=SIG_IGN, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1628d845a7}, NULL, 8) = 0 <
0.000004>
poll([{fd=10, events=POLLOUT}, {fd=7, events=POLLIN}], 2, 0) = 0 (Timeout) <0.000005>
rt_sigaction(SIGPIPE, NULL, {sa_handler=SIG_IGN, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1628d845a7}, 8) = 0 <
0.000004>
rt_sigaction(SIGPIPE, {sa_handler=SIG_IGN, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1628d845a7}, NULL, 8) = 0 <
0.000004>
poll([{fd=10, events=POLLPRI|POLLOUT|POLLWRNORM}], 1, 0) = 0 (Timeout) <0.000004>
rt_sigaction(SIGPIPE, {sa_handler=SIG_IGN, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1628d845a7}, NULL, 8) = 0 <
0.000003>
poll([{fd=10, events=POLLOUT}, {fd=7, events=POLLIN}], 2, 1000) = 0 (Timeout) <1.001341>
rt_sigaction(SIGPIPE, NULL, {sa_handler=SIG_IGN, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1628d845a7}, 8) = 0 <
0.000009>
rt_sigaction(SIGPIPE, {sa_handler=SIG_IGN, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1628d845a7}, NULL, 8) = 0 <
0.000010>
poll([{fd=10, events=POLLPRI|POLLOUT|POLLWRNORM}], 1, 0) = 0 (Timeout) <0.000006>
rt_sigaction(SIGPIPE, {sa_handler=SIG_IGN, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f1628d845a7}, NULL, 8) = 0 <
0.000004>
poll([{fd=10, events=POLLOUT}, {fd=7, events=POLLIN}], 2, 1000) = 0 (Timeout) <1.001340>
```

And it goes in a loop exactly 29 times. 192.168.35.1 is my HAProxy. Connection was blocked by a firewall, and it would try to connect 29 times with 1 second delay. So that's about ~30 seconds to run a command, and if it happen to try twice - that would be exact range 30-60 seconds to produce a line of logs.

That's the first issue. I am not sure `/var/www/html/apps/support` trying to connect to the external endpoint instead of internal container IP is a correct behavior or not. IMO for the traffic to go out to potentially public network only to return back... is unnecessary traffic at the very least. But I feel like something somewhere in the AIO should have alerted me that I have public endpoint port 443 unreachable...

And then... I have noticed that I'm having `--env SKIP_DOMAIN_VALIDATION=true` because... wait for it... it was complaining that it is unable to connect to 443. In the logs I had `internal or reserved ip-address error` and I do not remember much more details, but eventually I landed somewhere that suggested me to `SKIP_DOMAIN_VALIDATION`.

So - that is the second issue. I am a fool. Not sure you guys can fix that one.

The third issue is that I feel like, even though I had `SKIP_DOMAIN_VALIDATION` enabled, this issue had to come up somewhere in the logs if not on the main AIO page as a permanent warning. This would have saved a fool like me two sleepless nights. You have no idea... I was running Proxmox with ZFS and running AIO on LXC, and I thought that the issue I am having is due to https://github.com/nextcloud/all-in-one/discussions/1490. So, before I got my hands on `strace` - I nuked and re-installed entire Proxmox on LVM :,( Only to find out that the issue is still there..

**Comment by szaimen** [comment, marked as answer]
> That's the first issue. I am not sure `/var/www/html/apps/support` trying to connect to the external endpoint instead of internal container IP is a correct behavior or not. IMO for the traffic to go out to potentially public network only to return back... is unnecessary traffic at the very least. But I feel like something somewhere in the AIO should have alerted me that I have public endpoint port 443 unreachable...

It indeed needs to be able to connection the public domain via the public ip-address by default. If you dont want this see https://github.com/nextcloud/all-in-one#how-can-i-access-nextcloud-locally

> And then... I have noticed that I'm having `--env SKIP_DOMAIN_VALIDATION=true` because... wait for it... it was complaining that it is unable to connect to 443. In the logs I had `internal or reserved ip-address error` and I do not remember much more details, but eventually I landed somewhere that suggested me to `SKIP_DOMAIN_VALIDATION`.

Yes indeed the domain check showed that your network is not correctly configured

**Reply by dee-kryvenko** [reply]
Oh, I don't mind its ability to connect to it. I'm just thinking - it does it on every call which is just unnecessary traffic that could stay on the lo interface. Just the matter of optimization, that's all. I haven't measured how much traffic it produces, maybe negligible, just wanted to alert in case this wasn't expected behavior.

I see you converted this issue to discussion, does that mean you are not planning to address visibility of this issue? I truly believe that even with domain validation skip, there should be a warning in AIO or at the very least a visible record in the container logs. I understand that this is a user mistake, but it is extremely hard one to identify.

**Reply by szaimen** [reply]
> I truly believe that even with domain validation skip, there should be a warning in AIO or at the very least a visible record in the container logs. I understand that this is a user mistake, but it is extremely hard one to identify.

We have the domain validation for exactly this purpose.

**Reply by dee-kryvenko** [reply]
Which only happens once. It's not like I can't adjust my firewall rules after I install nextcloud, right?

**Reply by szaimen** [reply]
> It's not like I can't adjust my firewall rules after I install nextcloud, right?

Indeed you can still fix your firewall

**Reply by dee-kryvenko** [reply]
If I somehow know that it has to be fixed then yes. Otherwise I may just assume this is how slow NextCloud is and just go with something else.
